File: Design Pattern/Behavioral/Command/restaurant_order.ts
interface OrderCommand {
    execute(): void;
    undo(): void;
}

class Kitchen {
    prepareDish(dish: string): void {
        // TODO: Print "Preparing: " + dish
        console.log("Preparing: " + dish);
    }

    cancelDish(dish: string): void {
        // TODO: Print "Cancelling: " + dish
        console.log("Cancelling: "+ dish);
    }
}

class PlaceOrderCommand implements OrderCommand {
    private kitchen: Kitchen;
    private dish: string;

    constructor(kitchen: Kitchen, dish: string) {
        this.kitchen = kitchen;
        this.dish = dish;
    }

    execute(): void {
        // TODO: Call this.kitchen.prepareDish(this.dish)
        this.kitchen.prepareDish(this.dish)
    }

    undo(): void {
        // TODO: Call this.kitchen.cancelDish(this.dish)
        this.kitchen.cancelDish(this.dish)
    }
}

class CancelOrderCommand implements OrderCommand {
    private kitchen: Kitchen;
    private dish: string;

    constructor(kitchen: Kitchen, dish: string) {
        this.kitchen = kitchen;
        this.dish = dish;
    }

    execute(): void {
        // TODO: Call 
        this.kitchen.cancelDish(this.dish)
    }

    undo(): void {
        // TODO: Call this.kitchen.prepareDish(this.dish)
        this.kitchen.prepareDish(this.dish)
    }
}

class Waiter {
    private pending: OrderCommand[] = [];
    private history: OrderCommand[] = [];

    takeOrder(command: OrderCommand): void {
        // TODO: Add command to pending queue
        this.pending.push(command);
    }

    submitOrders(): void {
        // TODO: Execute all pending commands, move to history, clear pending
        for(const cmd of this.pending) {
            cmd.execute();
            this.history.push(cmd);
        }
        this.pending = [];
    }

    undoLast(): void {
        // TODO: Pop from history and call undo()
        if (this.history.length > 0) {
            this.history.pop()!.undo();
        }
    }
}

const kitchen = new Kitchen();
const waiter = new Waiter();
waiter.takeOrder(new PlaceOrderCommand(kitchen, "Pasta"));
waiter.takeOrder(new PlaceOrderCommand(kitchen, "Salad"));
waiter.submitOrders();
waiter.takeOrder(new CancelOrderCommand(kitchen, "Salad"));
waiter.submitOrders();
waiter.undoLast();

File: Design Pattern/Behavioral/Command/restaurant_order.cpp
#include <iostream>
#include <queue>
#include <stack>
#include <string>
using namespace std;

class OrderCommand {
public:
  virtual ~OrderCommand() = default;
  virtual void execute() = 0;
  virtual void undo() = 0;
};

class Kitchen {
public:
  void prepareDish(const string &dish) {
    // TODO: Print "Preparing: [dish]"
    cout << "Preparing: " << dish << endl;
  }

  void cancelDish(const string &dish) {
    // TODO: Print "Cancelling: [dish]"
    cout << "Cancelling: " << dish << endl;
  }
};

class PlaceOrderCommand : public OrderCommand {
  // TODO: Add fields (kitchen, dish)
  Kitchen *kitchen;
  string dish;

public:
  PlaceOrderCommand(Kitchen *kitchen, const string &dish) {
    // TODO: Store receiver and dish
    this->kitchen = kitchen;
    this->dish = dish;
  }

  void execute() override {
    // TODO: Call kitchen->prepareDish(dish)
    kitchen->prepareDish(dish);
  }

  void undo() override {
    // TODO: Call kitchen->cancelDish(dish)
    kitchen->cancelDish(dish);
  }
};

class CancelOrderCommand : public OrderCommand {
  // TODO: Add fields (kitchen, dish)
  Kitchen *kitchen;
  string dish;

public:
  CancelOrderCommand(Kitchen *kitchen, const string &dish) {
    // TODO: Store receiver and dish
    this->kitchen = kitchen;
    this->dish = dish;
  }

  void execute() override {
    // TODO: Call kitchen->cancelDish(dish)
    kitchen->cancelDish(dish);
  }

  void undo() override {
    // TODO: Call kitchen->prepareDish(dish)
    kitchen->prepareDish(dish);
  }
};

class Waiter {
  // TODO: Add a pending queue and a history stack
  queue<OrderCommand *> pending;
  stack<OrderCommand *> history;

public:
  void takeOrder(OrderCommand *command) {
    // TODO: Add command to pending queue
    pending.push(command);
  }

  void submitOrders() {
    // TODO: Execute all pending commands, move them to history
    while (!pending.empty()) {
      auto front = pending.front();
      pending.pop();
      front->execute();
      history.push(front);
    }
  }

  void undoLast() {
    // TODO: Pop the most recent command from history and call undo()
    auto top = history.top();
    history.pop();
    top->undo();
  }
};

int main() {
  Kitchen kitchen;
  Waiter waiter;
  waiter.takeOrder(new PlaceOrderCommand(&kitchen, "Pasta"));
  waiter.takeOrder(new PlaceOrderCommand(&kitchen, "Salad"));
  waiter.submitOrders();
  waiter.takeOrder(new CancelOrderCommand(&kitchen, "Salad"));
  waiter.submitOrders();
  waiter.undoLast();
  return 0;
}
